src/platform/macos/events_loop.rs

```diff
@@ -544,6 +544,67 @@ impl Proxy {
     }
 }
 
+pub fn windows_key_code_for_char_code(c: char) -> Option<c_ushort> {
+    // We only translate keys that are affected by keyboard layout.
+    //
+    // Note that since keys are translated in a somewhat "dumb" way (reading character)
+    // there is a concern that some combination, i.e. Cmd+char, causes the wrong
+    // letter to be received, and so we receive the wrong key.
+    //
+    // Codes taken from to_virtual_key_code.
+    // It may be better to map these codes to something more immediately meaningful.
+    Some(match c {
+        'a' | 'A' => 0x00,
+        's' | 'S' => 0x01,
+        'd' | 'D' => 0x02,
+        'f' | 'F' => 0x03,
+        'h' | 'H' => 0x04,
+        'g' | 'G' => 0x05,
+        'z' | 'Z' => 0x06,
+        'x' | 'X' => 0x07,
+        'c' | 'C' => 0x08,
+        'v' | 'V' => 0x09,
+        'b' | 'B' => 0x0b,
+        'q' | 'Q' => 0x0c,
+        'w' | 'W' => 0x0d,
+        'e' | 'E' => 0x0e,
+        'r' | 'R' => 0x0f,
+        'y' | 'Y' => 0x10,
+        't' | 'T' => 0x11,
+        '1' | '!' => 0x12,
+        '2' | '@' => 0x13,
+        '3' | '#' => 0x14,
+        '4' | '$' => 0x15,
+        '6' | '^' => 0x16,
+        '5' | '%' => 0x17,
+        '=' | '+' => 0x18,
+        '9' | '(' => 0x19,
+        '7' | '&' => 0x1a,
+        '-' | '_' => 0x1b,
+        '8' | '*' => 0x1c,
+        '0' | ')' => 0x1d,
+        ']' | '}' => 0x1e,
+        'o' | 'O' => 0x1f,
+        'u' | 'U' => 0x20,
+        '[' | '{' => 0x21,
+        'i' | 'I' => 0x22,
+        'p' | 'P' => 0x23,
+        'l' | 'L' => 0x25,
+        'j' | 'J' => 0x26,
+        '\''| '"' => 0x27,
+        'k' | 'K' => 0x28,
+        ';' | ':' => 0x29,
+        '\\'| '|' => 0x2a,
+        ',' | '<' => 0x2b,
+        '/' | '?' => 0x2c,
+        'n' | 'N' => 0x2d,
+        'm' | 'M' => 0x2e,
+        '.' | '>' => 0x2f,
+        '`' | '~' => 0x32,
+        _ => return None,
+    })
+}
+
 pub fn to_virtual_key_code(code: c_ushort) -> Option<events::VirtualKeyCode> {
```

Comment:
```suggestion
pub fn scancode_to_keycode(code: c_ushort) -> Option<events::VirtualKeyCode> {
```
It would be good to give this function a name that makes the difference explicit between it and the char-based function. Once again, feel free to try a different naming pattern than what I suggested.

`check_additional_virtual_key_codes` should probably also be renamed to be 1) consistent with whatever names the other ones will have, and 2) make the growing number of keycode functions less confusing... that could just be `check_function_keys`, or something like that.

Reply:
Would it make sense to remove the `check_additional_virtual_key_codes()` function and just move its checks to `char_to_keycode()`?  This would avoid doubling up on string checks (both functions take virtually the same input) but loses the fact that `char_to_keycode()` only checks layout related keys.

I don't have a keyboard with those function keys to test anything, and it's probably not worthwhile, so I'll leave that TBD.

Reply:
The only complication is that the function keys use UTF-16, but I'm definitely in favor of merging them ***if*** it makes things cleaner.

I'm not really able to test those keys either, and neither was the person who implemented it IIRC. It's straightforward enough that I'm not too worried, though.

Reply:
When I press my F keys I get characters like `"\u{f703}"`.  It seems like we could match against these for those keys and avoid UTF-16.  Or is there a specific reason for the larger encoding size?

Reply:
The fact that the docs are explicit about this: https://developer.apple.com/documentation/appkit/1535851-function-key_unicodes?preferredLanguage=occ

Reply:
I've decided to leave `check_additional_virtual_key_codes()` (now `check_function_keys()`) as it is for now, minus the change from `Option<&String>` to `&String` to align with `get_characters()`.  

I don't think it would make things any cleaner and makes the `char_to_keycode()` code less purposeful (imo).